### PR TITLE
Avoid grad_fn getting garbage collected in multihooks

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102175

Workaround for https://github.com/pytorch/pytorch/issues/102174

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -391,8 +392,9 @@ def inner_hook(grad: torch.Tensor):
     class Handle(RemovableHandle):
         handles: Tuple[RemovableHandle, ...]
 
-        def __init__(self, handles: Tuple[RemovableHandle, ...]):
+        def __init__(self, handles: Tuple[RemovableHandle, ...], grad_fns: Tuple[Node, ...]):
             self.handles = handles
+            self.grad_fns: Optional[Tuple] = grad_fns
```
Comment (awgu) [inline torch/autograd/graph.py:397]: Regarding `Optional[Tuple]` typing, when can `grad_fns` be `None`?

**Review by awgu — APPROVED** [review]
This makes sense to me! Feel free to wait for @albanD's review.

File: torch/autograd/graph.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/graph.py:397]: Whoops I meant to set it to None in `remove`, good catch

**Comment by awgu** [comment]
Hi @albanD. Would it be possible to get a review on this? I would like to add land the improved fine-tuning support for FSDP (https://github.com/pytorch/pytorch/pull/101982/) that depends on this.

**Comment by soulitzer** [comment]
@awgu this is probably fine, but from discussion with Alban, we probably just want to fix the underlying issue. I added a comment on the issue.

**Comment by awgu** [comment]
> @awgu this is probably correct, but from discussion with Alban, we probably just want to fix the underlying issue (should be a one line fix). I added a comment on the issue.

Oh okay, so will we still land this PR or will there be a separate PR with the proper fix?

**Comment by soulitzer** [comment]
Yeah separate PR, closing this one

**Pull request closed without merging**